From d84a0ba37df25e48de1c5c1533a4e50d826f010f Mon Sep 17 00:00:00 2001
From: Stanislav Kinsburskii <skinsburskii@microsoft.com>
Date: Wed, 28 Aug 2024 16:37:02 +0000
Subject: [PATCH] Merged PR 167: heki: Add ability to skip frame protection

```
heki: Add ability to skip frame protection

By default, heki tries to protect all the existent pages within the range
provided. This might be not suitable in some cases.
Provide a weak function, allowing to skip page protection if overwritten.

Signed-off-by: Stanislav Kinsburskii <skinsburskii@linux.microsoft.com>
```

----
#### AI description  (iteration 1)
#### PR Classification
New feature: Add ability to skip frame protection.

#### PR Summary
This pull request introduces a new feature that allows skipping frame protection based on a condition.
- `virt/heki/protect.c`: Added a weak function `heki_protect_pfn` to determine if a page frame number (pfn) should be protected.
- `virt/heki/protect.c`: Updated `heki_protect_cb` to skip protection for pfns that do not meet the condition defined in `heki_protect_pfn`.
---
 virt/heki/protect.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/virt/heki/protect.c b/virt/heki/protect.c
index 8d8db3ed282e..947be4af5ec3 100644
--- a/virt/heki/protect.c
+++ b/virt/heki/protect.c
@@ -46,6 +46,11 @@ void heki_init_perm(unsigned long va, unsigned long end, struct heki_args *args)
 	heki_walk(va, end, heki_init_perm_cb, args);
 }
 
+bool __weak heki_protect_pfn(unsigned long pfn)
+{
+	return true;
+}
+
 static void heki_protect_cb(struct heki_args *args)
 {
 	unsigned long va;
@@ -63,6 +68,9 @@ static void heki_protect_cb(struct heki_args *args)
 	     pa += PAGE_SIZE, va += PAGE_SIZE) {
 
 		pfn = pa >> PAGE_SHIFT;
+		if (!heki_protect_pfn(pfn))
+			continue;
+
 		cur_perm = (unsigned long) xa_load(&args->permissions, pfn);
 
 		args->attributes = cur_perm | perm;
-- 
2.42.0

